pkg/controllers/workgenerator/controller.go

```diff
@@ -189,35 +223,47 @@ func (r *Reconciler) Reconcile(ctx context.Context, req controllerruntime.Reques
 				Message:            "In the processing of synchronizing the work to the member cluster",
 				ObservedGeneration: resourceBinding.Generation,
 			})
-		} else {
-			setBindingStatus(works, &resourceBinding)
+			resourceBinding.SetConditions(metav1.Condition{
+				Status:             metav1.ConditionFalse,
+				Type:               string(fleetv1beta1.ResourceBindingAvailable),
+				Reason:             condition.WorkNeedSyncedReason,
+				Message:            "In the processing of synchronizing the work to the member cluster",
+				ObservedGeneration: resourceBinding.Generation,
+			})
+			klog.V(2).InfoS("Work Updated. Updating the resourceBinding", "resourceBinding", klog.KObj(resourceBinding), "resourceBindingStatus", resourceBinding.Status)
 		}
 	}
+}
 
-	// update the resource binding status
+func (r *Reconciler) updateBinding(ctx context.Context, resourceBinding fleetv1beta1.ClusterResourceBinding, overrideSucceeded, workUpdated bool, syncErr error) error {
 	if updateErr := r.Client.Status().Update(ctx, &resourceBinding); updateErr != nil {
-		klog.ErrorS(updateErr, "Failed to update the resourceBinding status", "resourceBinding", bindingRef)
-		return controllerruntime.Result{}, controller.NewUpdateIgnoreConflictError(updateErr)
-	}
-	if errors.Is(syncErr, controller.ErrUserError) {
-		// Stop retry when the error is caused by user error
-		// For example, user provides an invalid overrides or cannot extract the resources from config map.
-		klog.ErrorS(syncErr, "Stopped retrying the resource binding", "resourceBinding", bindingRef)
-		return controllerruntime.Result{}, nil
-	}
-
-	if errors.Is(syncErr, errResourceSnapshotNotFound) {
-		// This error usually indicates that the resource snapshot is deleted since the rollout controller which fills
-		// the resource snapshot share the same informer cache with this controller. We don't need to retry in this case
-		// since the resource snapshot will not come back. We will get another event if the binding is pointing to a new resource.
-		// However, this error can happen when the resource snapshot exists during the IT test when the client that creates
-		// the resource snapshot is not the same as the controller client so that we need to retry in this case.
-		// This error can also happen if the user uses a customized rollout controller that does not share the same informer cache with this controller.
-		return controllerruntime.Result{Requeue: true}, nil
+		klog.ErrorS(updateErr, "Failed to update the resourceBinding status", "resourceBinding", resourceBinding.Status)
+		// Retry only for specific errors or conditions
+		errAfterRetries := retry.RetryOnConflict(retry.DefaultRetry, func() error {
```

Comment:
I also realized that we can't blindly retry this as the binding spec may have changed. I think the safer way is to delete the available condition if the applied Condition is false. Worst case is that we might have to add a "lastUpdatedTime" field to our conditions.

Reply:
we should only update the sync/override/applied/available condition on the binding status which is safe.